README.md
Kent's Emulated GS Apple //GS Emulator

(Git repositoy with the blessing of Kent Dickey)

This is a git repository of all of the releases of Kent's 
Emulated GS after 0.30.  Previous to 0.30, the code was
called "sim65" and was HP-PA assembly only, and is not 
included here.

Kent's sourceforge site can be found here:
[Kent's Emulated  GS](http://kegs.sourceforge.net/)

Kent was kind enough to provide all of the tar's for the
releases under the condition that I release publicly any
script that I used for importing the data.  The script
I created is a total hack but can be found in the tools 
directory (with a nod to the FTA -- "Winners don't use tools") 
It should be wholely uninteresting given that you are currently 
looking at the resulting git repository.

I will attempt to keep the main repository up to date
with the latest from Kent, including this README file.
Each release that Kent makes is tagged with the contents 
of his tar.gz release, and the tag'd versions will not contain 
any extras (like this file or the tools directory).





CHANGES
Changes in KEGS v0.63 since v0.62 (08/15/02) <frederic.devernay@m4x.org>
- Added F6 function, which defaults to "enter debugger" (for 1-button mouse)
- F12 now defaults to "switch speed limit" (for 1-button mouse)
- Command-line help ('kegs -\?' or 'kegs -h')
- Debugger help ('?')
- make_size script should work with any perl installation
- Config files can be found in the resources folder on MacOS X
  (KEGS.app/Contents/Resources)
- Control-backspace resets KEGS (for MacOS users who don't have a Pause key)

Changes in KEGS v0.62 since v0.61 (06/22/02) <frederic.devernay@m4x.org>
- Slow paddles mode (<inwards@inwards.com>)
- Fixed pointer warping (F8 by default) <inwards@inwards.com>

Changes in KEGS v0.61 since v0.60 (02/19/02) <frederic.devernay@m4x.org>
- SDL video/sound/joystick driver, see www.libsdl.org
- code cleanups (prototypes, etc.)

Changes in KEGS v0.60 since v0.59 (10/03/00)
- The 16-bit colors were still wrong due to another coding error.  It would
	be much easier to get this right if I had a 16-bit color display...
	A user says it works now.

Changes in KEGS v0.59 since v0.58 (7/07/00)
- Added support for multiple paths to the default files and also multiple
	names for many default files.  This should make the .rpm distribution
	work better.
- Add another keycode to mean break according to mic@research.nj.nec.com.
- Add support for various ROMs to get plugged into slot 1-7.
- Fix code so that it should compile on 64-bit platforms.

Changes in KEGS v0.58 since v0.57 (2/08/00)
- Setting the execute bit on the disk image no longer means no-write-thru.
	Too many new users were getting confused by this.
- Fixed another bug with Apple //e bank switching created by v0.56
	Reported by phoenyx.
- Add command line option "-v" to turn on some verbose debugging flags.
- Fixed potential core-dump bug with non-8 bit visuals.
- Fixed double-lo-res color problem.
- The X driver should work with any visual depth display now and get the
	colors right.  Ian Schmidt reported his 16-bit card had bad colors.

Changes in KEGS v0.57 since v0.56 (12/27/99)
- Another try at making timezone stuff work across all Unix variants.
	Let me know if the time in the Apple //gs control panel doesn't
	match your real local time.
- Fix a bug created in v0.56 where the fast //e bank switch code had a typo.
	This prevented ZBasic from working correctly.

Changes in KEGS v0.56 since v0.55 (10/31/99)
- Faster Apple //e bank switch emulation.
- Simplified number of global variables for various softswitches.
- Fixed a bug which made 3.5" and 5.25" disk access much slower than necessary.
- Improved scan-line interrupt accuracy (lets MEGADEMO run).
- Improved sound interrupt accuracy (was hoping this would fix some sound
	issues, but it doesn't seem to help).
- Add Mode_switch as an alias for the Option key
- I noticed the //gs self-tests were broken again--fixed.
 
Changes in KEGS v0.55 since v0.54 (10/19/99)
- In LOG_PC debug aid, add cycles to the trace
- Fix MEGADEMO bug where 3.5" disks weren't properly ejected.  Needed to
	look at iwm.motor_on35 not iwm.motor_on.
- Temp fix for MEGADEMO to not halt if shadow-in-all-banks is on in $c036.
- Another MEGADEMO fix to not take a scan-line int if the SCB was cleared
	right before the raster got to this line.
- Fix bug in smartport.c that was causing core dumps if you tried to init
	a disk is s7dx.

Changes in KEGS v0.54 since v0.53 (10/10/99)
- Add support for Out Of This World's direct reading of ADB RAM loc 0xb to
	get key status.  This lets shift/control work in OOTW.
- Code simplification to get rid of most set_halt() calls and use halt_printf.
- Speed improvement: track kpc (merged kbank and pc in one 32 bit variable)
	which makes the inner loop faster.  This does make KEGS not
	accurately model a 65816 code crossing bank boundaries, but just
	about every other emulator gets it wrong, and the speed improvement
	is 5-10%.  And I don't know of any code which relies on it
	working correctly.
- Fix to allow better GS/OS compatibility: after each smartport call,
	set 0x7f8 = 0xc7.
- Fixed ZipGS emulation bug where KEGS was not re-locking Zip at the right
	time, which made double-hires not work after booting GS/OS.

Changes in KEGS v0.53 since v0.52 (8/3/99)
- Move all the "fcycles" timing calculations to use double instead of float.
- Fix display shadowing bug reported by "phoenyx" which caused the text
	display to not always be updated correctly with funny bank switching.
- Added the "Home" key as an alias for the '=' on the keypad.
- Changed the way X modifiers are interpreted to increase compatibility of
	Caps Lock to more X servers.
- Add -dhr140 option to use old double-hires color mode that results in
	exactly 140 horizontal pixels with no bleeding.  It's set default
	to "on" for now while I work out double-hires colors.
- Started adding some ZipGS compatibility--control panels run, but all
	the controls are effectively ignored by KEGS.

Changes in KEGS v0.52 since v0.51 (6/27/99)
- Small speed-up of interpreter loop to avoid checking the global variable
	"halt_sim" after every instruction.
- Smartport fixes to avoid halts when the SCSI CD player NDA is installed.
- Fix to autodetect X visual depth (it didn't work at all in v0.51).
- Fix to HP binary--KEGS v0.51 hit an HP linker bug which caused the
	executable to not run correctly.  (It didn't obey an assembly-
	language alignment command correctly).  Re-ordering the object
	list works around the problem.

Changes in KEGS v0.51 since v0.50 (6/1/99)
- Fixed many bugs that crept into scanline interrupts over the last few months.
- RAM size is now settable on the commandline: -mem 0x400000 will use
	a 4MB expansion RAM card (giving you 4.25MB of memory with ROM 01).
- VBL time used to be a variable (which was wrong)--it's now always the
	same number of cycles.
- Typo preventing joystick_driver.c from compiling fixed.
- Auto senses X visual depth, searching for 8 bit, then 15 bit, then 24,

src/woz.c
const char rcsid_woz_c[] = "@(#)$KmKId: woz.c,v 1.7 2021-05-04 04:43:08+00 kentd Exp $";

/************************************************************************/
/*			KEGS: Apple //gs Emulator			*/
/*			Copyright 2002-2021 by Kent Dickey		*/
/*									*/
/*	This code is covered by the GNU GPL v3				*/
/*	See the file COPYING.txt or https://www.gnu.org/licenses/	*/
/*	This program is provided with no warranty			*/
/*									*/
/*	The KEGS web page is kegs.sourceforge.net			*/
/*	You may contact the author at: kadickey@alumni.princeton.edu	*/
/************************************************************************/

// Based on WOZ 2.0/1.0 spec at https://applesaucefdc.com/woz/reference2/

#include "defc.h"

extern const char g_kegs_version_str[];

byte g_woz_hdr_bytes[8] = { 0x57, 0x4f, 0x5a, 0x32,		// "WOZ2"
				0xff, 0x0a, 0x0d, 0x0a };

void
woz_parse_meta(Disk *dsk, byte *bptr, int size)
{
	Woz_info *wozinfo_ptr;
	int	c;
	int	i;

	wozinfo_ptr = dsk->wozinfo_ptr;
	if(wozinfo_ptr->meta_bptr) {
		printf("Bad WOZ file, 2 META chunks\n");
		wozinfo_ptr->woz_size = 0;
		return;
	}
	wozinfo_ptr->meta_bptr = bptr;
	wozinfo_ptr->meta_size = size;
	printf("META field, %d bytes:\n", size);
	for(i = 0; i < size; i++) {
		c = bptr[i];
		if(c == 0) {
			break;
		}
		putchar(c);
	}
	putchar('\n');
}

void
woz_parse_info(Disk *dsk, byte *bptr, int size)
{
	Woz_info *wozinfo_ptr;
	byte	*new_str;
	int	info_version, disk_type, write_protect, synchronized;
	int	cleaned, ram, largest_track;

	wozinfo_ptr = dsk->wozinfo_ptr;
	if(wozinfo_ptr->info_bptr) {
		printf("Two INFO chunks, bad WOZ file\n");
		wozinfo_ptr->woz_size = 0;
		return;
	}
	wozinfo_ptr->info_bptr = bptr;
	if(size < 60) {
		printf("INFO field is %d, too short\n", size);
		wozinfo_ptr->woz_size = 0;
		return;
	}
	info_version = bptr[0];		// Only "1" or "2" is supported
	disk_type = bptr[1];		// 1==5.25", 2=3.5"
	write_protect = bptr[2];	// 1==write protected
	synchronized = bptr[3];		// 1==cross track sync during imaging
	cleaned = bptr[4];		// 1==MC3470 fake bits have been removed
	new_str = malloc(32 + 1);
	memcpy(new_str, &(bptr[5]), 32);
	new_str[32] = 0;		// Null terminate
	printf("INFO, %d bytes.  info_version:%d, disk_type:%d, wp:%d, sync:"
		"%d, cleaned:%d\n", size, info_version, disk_type,
		write_protect, synchronized, cleaned);
	printf("Creator: %s\n", new_str);
	free(new_str);
	if(info_version >= 2) {
		ram = bptr[42] + (bptr[43] << 8);
		largest_track = (bptr[44] + (bptr[45] << 8)) * 512;
		printf("Disk sides:%d, boot_format:%d bit_timing:%d, hw:"
			"%02x%02x, ram:%d, largest_track:0x%07x\n", bptr[37],
			bptr[38], bptr[39], bptr[41], bptr[40], ram,
			largest_track);
	}

	if(write_protect) {
		printf("Write protected\n");
		dsk->write_prot = 1;
		dsk->write_through_to_unix = 0;
	}
}

void
woz_parse_tmap(Disk *dsk, byte *bptr, int size)
{
	Woz_info *wozinfo_ptr;
	int	i;

	wozinfo_ptr = dsk->wozinfo_ptr;
	if(wozinfo_ptr->tmap_bptr) {
		printf("Second TMAP chunk, bad WOZ file!\n");
		wozinfo_ptr->woz_size = 0;
		return;
	}
	wozinfo_ptr->tmap_bptr = bptr;
	printf("TMAP field, %d bytes\n", size);
	for(i = 0; i < 36; i++) {
		printf("Track %2d.00: %02x, %2d.25:%02x %2d.50:%02x %2d.75:"
			"%02x\n", i, bptr[0], i, bptr[1],
			i, bptr[2], i, bptr[3]);
		bptr += 4;
	}
	size -= (36*4);
	for(i = 0; i < size; i++) {
		if(bptr[i] != 0xff) {
			printf("TMAP entry at offset +0x%03x is %02x\n",
					36*4 + i, bptr[i]);
		}
	}
}

void
woz_parse_trks(Disk *dsk, byte *bptr, int size)
{
	Woz_info *wozinfo_ptr;

	printf("TRKS field, %d bytes\n", size);
	wozinfo_ptr = dsk->wozinfo_ptr;
	if(wozinfo_ptr->trks_bptr) {
		printf("Second TRKS chunk, illegal Woz file\n");
		wozinfo_ptr->woz_size = 0;
		return;
	}
	wozinfo_ptr->trks_bptr = bptr;
	wozinfo_ptr->trks_size = size;
}

int
woz_add_track(Disk *dsk, int qtr_track, word32 tmap, double dcycs)
{
	Woz_info *wozinfo_ptr;
	Trk	*trk;
	byte	*trks_bptr, *raw_bptr, *sync_ptr, *bptr;
	word32	raw_bytes, num_bytes, trks_size, len_bits, offset, num_blocks;
	word32	block;
	int	i;

	wozinfo_ptr = dsk->wozinfo_ptr;
	trks_bptr = wozinfo_ptr->trks_bptr;
	trks_size = wozinfo_ptr->trks_size;
	if(wozinfo_ptr->version == 1) {
		offset = tmap * 6656;
		if((offset + 6656) > trks_size) {
			printf("Trk %d is out of range!\n", tmap);
			return 0;
		}

		bptr = trks_bptr + offset;
		len_bits = bptr[6648] | (bptr[6649] << 8);
		if(len_bits > (6656*8)) {
			printf("Trk bits: %d too big\n", len_bits);
			return 0;
		}
	} else {
		bptr = &(trks_bptr[tmap * 8]);
		// This is a TRK 8-byte structure
		block = cfg_get_le16(&bptr[0]);		// Starting Block
		num_blocks = cfg_get_le16(&bptr[2]);	// Block Count
		len_bits = cfg_get_le32(&bptr[4]);	// Bits Count
#if 0
		printf("qtr_track:%02x, block:%04x, num_blocks:%04x, "
			"len_bits:%06x\n", qtr_track, block, num_blocks,
			len_bits);
		printf("File offset: %05lx\n", bptr - wozinfo_ptr->wozptr);
#endif

		if(block < 3) {
			return 0;
		}
		offset = (block - 3) * 512;		// Offset from +1536
		offset = offset + 1536 - 256;		// Offset from trks_bptr
		if((offset + (num_blocks * 512)) > trks_size) {
			printf("Trk %d is out of range!\n", tmap);
			return 0;
		}
		bptr = &(trks_bptr[offset]);
#if 0
		printf("Qtr_track %03x offset:%06x, bptr:%p, trks_bptr:%p\n",
			qtr_track, offset, bptr, trks_bptr);
		printf(" len_bits:%d %06x bptr-wozptr: %07lx\n", len_bits,
				len_bits, bptr - wozinfo_ptr->wozptr);
#endif
		if(len_bits > (num_blocks * 512 * 8)) {
			printf("Trk bits: %d too big\n", len_bits);
			return 0;
		}
	}

	raw_bytes = (len_bits + 7) >> 3;
	num_bytes = 2 + raw_bytes + 4;
	trk = &(dsk->trks[qtr_track]);
	trk->track_qbits = len_bits * 4;
	trk->dunix_pos = 256 + offset;
	trk->unix_len = raw_bytes;
	trk->raw_bptr = (byte *)malloc(num_bytes);
	trk->sync_ptr = (byte *)malloc(num_bytes);

	iwm_move_to_track(dsk, qtr_track);
	raw_bptr = &(trk->raw_bptr[2]);
	sync_ptr = &(trk->sync_ptr[2]);
	for(i = 0; i < raw_bytes; i++) {
		raw_bptr[i] = bptr[i];
		sync_ptr[i] = 0xff;
	}

	iwm_recalc_sync(dsk, 0, dcycs);

	if(qtr_track == 0) {
		printf("Track 0 data begins: %02x %02x %02x, offset:%d\n",
			bptr[0], bptr[1], bptr[2], offset);
	}

	return 1;
}

int
woz_parse_header(Disk *dsk, byte *wozptr, word32 woz_size, double dcycs)
{
	Woz_info *wozinfo_ptr;
	byte	*bptr, *tmap_bptr;
	word32	chunk_id, size, tmap;
	int	pos, version, ret, num_tracks;
	int	i;

	wozinfo_ptr = dsk->wozinfo_ptr;
	version = 2;
	if(woz_size < 8) {
		return 0;
	}
	for(i = 0; i < 8; i++) {
		if(wozptr[i] != g_woz_hdr_bytes[i]) {
			if(i == 3) {		// Check for WOZ1
				if(wozptr[i] == 0x31) {		// WOZ1
					version = 1;
					continue;
				}
			}
			printf("WOZ header[%d]=%02x, invalid\n", i, wozptr[i]);
			return 0;
		}
	}
	wozinfo_ptr->version = version;

	printf("WOZ version: %d\n", version);
	pos = 12;
	while(pos < woz_size) {
		bptr = &(wozptr[pos]);
		chunk_id = bptr[0] | (bptr[1] << 8) | (bptr[2] << 16) |
				(bptr[3] << 24);
		size = bptr[4] | (bptr[5] << 8) | (bptr[6] << 16) |
							(bptr[7] << 24);
		pos += 8;
		printf("chunk_id: %08x, size:%08x\n", chunk_id, size);
		if(((pos + size) > woz_size) || (size < 8) ||
							((size >> 30) != 0)) {
			return 0;
		}
		bptr = &(wozptr[pos]);
		if(chunk_id == 0x4f464e49) {		// "INFO"
			woz_parse_info(dsk, bptr, size);
		} else if(chunk_id == 0x50414d54) {	// "TMAP"
			woz_parse_tmap(dsk, bptr, size);
		} else if(chunk_id == 0x534b5254) {	// "TRKS"
			woz_parse_trks(dsk, bptr, size);
		} else if(chunk_id == 0x4154454d) {	// "META"
			woz_parse_meta(dsk, bptr, size);
		} else {
			printf("Chunk header %08x is unknown\n", chunk_id);
		}
		pos += size;
	}

	if(!wozinfo_ptr->tmap_bptr) {
		printf("No TMAP found\n");
		return 0;
	}
	if(!wozinfo_ptr->trks_bptr) {
		printf("No TRKS found\n");
		return 0;
	}
	if(!wozinfo_ptr->info_bptr) {
		printf("No INFO found\n");
		return 0;
	}
	if(wozinfo_ptr->woz_size == 0) {
		return 0;
	}

	tmap_bptr = wozinfo_ptr->tmap_bptr;
	dsk->cur_qbit_pos = 0;
	num_tracks = 4*35;
	if(!dsk->disk_525) {
		num_tracks = 160;
	}
	disk_set_num_tracks(dsk, num_tracks);
	for(i = 0; i < 160; i++) {
		if(dsk->disk_525) {
			if(i & 1) {		// qtr-track
				continue;	// Not supported yet
			}
		}
		tmap = tmap_bptr[i];
		if(tmap >= 0xff) {
			continue;		// Skip
		}
		ret = woz_add_track(dsk, i, tmap, dcycs);
		if(ret == 0) {
			return ret;
		}
	}

	return 1;		// WOZ file is good!
}

int
woz_open(Disk *dsk, double dcycs)
{
	Woz_info *wozinfo_ptr;
	byte	*wozptr;
	word32	woz_size, off;
	int	fd, ret;

	// return 0 for bad WOZ file, 1 for success
	printf("woz_open on file %s, write_prot:%d\n", dsk->name_ptr,
						dsk->write_prot);
	if(dsk->raw_data) {
		wozptr = dsk->raw_data;
		woz_size = dsk->raw_dsize;
		dsk->write_prot = 1;
		dsk->write_through_to_unix = 0;
	} else {
		fd = dsk->fd;
		if(fd < 0) {
			return 0;
		}
		woz_size = cfg_get_fd_size(fd);
		printf("size: %d\n", woz_size);

		wozptr = malloc(woz_size);
		off = cfg_read_from_fd(fd, wozptr, 0, woz_size);
		if(off != woz_size) {
			close(fd);
			return 0;
		}
	}

	wozinfo_ptr = malloc(sizeof(Woz_info));
	if(!wozinfo_ptr) {
		fprintf(stderr, "Out of memory\n");
		exit(1);
	}

	dsk->wozinfo_ptr = wozinfo_ptr;

	wozinfo_ptr->wozptr = wozptr;
	wozinfo_ptr->woz_size = woz_size;
	wozinfo_ptr->version = 0;
	wozinfo_ptr->bad = 0;
	wozinfo_ptr->tmap_bptr = 0;
	wozinfo_ptr->trks_bptr = 0;
	wozinfo_ptr->info_bptr = 0;
	wozinfo_ptr->meta_bptr = 0;
	wozinfo_ptr->meta_size = 0;

	ret = woz_parse_header(dsk, wozptr, woz_size, dcycs);
	printf("woz_parse_header ret:%d, write_prot:%d\n", ret,
							dsk->write_prot);
	return ret;
}

int
woz_new(int fd, const char *str, int size_kb)
{
	byte	buf[0x600];
	word32	size, type;
	int	c;
	int	i;

	// Write to fd (set to byte 0 of the file) to write new WOZ file
	memset(&buf[0], 0, 0x600);
	memcpy(&buf[0], &g_woz_hdr_bytes[0], 8);	// "WOZ2", other stuff

	// INFO
	cfg_set_le32(&buf[12], 0x4f464e49);		// "INFO"
	buf[16] = 60;					// INFO length
	buf[20] = 2;					// WOZ2
	type = 1;					// 5.25"
	if(size_kb > 140) {
		type = 2;				// 3.5"
	}
	buf[21] = type;
	for(i = 0; i < 32; i++) {
		buf[25 + i] = ' ';			// Creator field
	}
	memcpy(&buf[25], "KEGS", 4);
	for(i = 0; i < 20; i++) {
		c = g_kegs_version_str[i];
		if(c == 0) {
			break;
		}
		buf[25 + 5 + i] = c;
	}
	buf[57] = type;					// sides, re-use type
	buf[59] = 32;

	// TMAP
	cfg_set_le32(&buf[80], 0x50414d54);		// TMAP
	buf[84] = 160;
	for(i = 0; i < 160; i++) {
		buf[88 + i] = 0xff;
	}

	// TRKS
	cfg_set_le32(&buf[248], 0x534b5254);		// TRKS
	cfg_set_le32(&buf[252], 1280);			// Room for array only

	size = must_write(fd, &buf[0], 0x600);
	if(size != 0x600) {
		return -1;
	}

	return 0;
}
